File: CON BDD/EUREKABANK_SOAP_JAVA_GR06/08. SERVIDOR/eurekabank_soap_java_gr06/src/main/java/ec/edu/monster/persistencia/TasaCambioDAO.java
package ec.edu.monster.persistencia;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;

/** Acceso a la tabla `tasacambio`. */
public class TasaCambioDAO {

    /**
     * Tasa para convertir un monto de la moneda {@code origen} a {@code destino}.
     * Si son iguales devuelve 1. Debe ejecutarse dentro de la transacción.
     *
     * @throws SQLException si no existe la tasa configurada.
     */
    public double tasa(Connection cn, String origen, String destino) throws SQLException {
        if (origen == null || destino == null || origen.equals(destino)) {
            return 1.0d;
        }
        String sql = "SELECT dec_tasa FROM tasacambio "
                   + "WHERE chr_origen = ? AND chr_destino = ?";
        try (PreparedStatement ps = cn.prepareStatement(sql)) {
            ps.setString(1, origen);
            ps.setString(2, destino);
            try (ResultSet rs = ps.executeQuery()) {
                if (rs.next()) {
                    return rs.getDouble("dec_tasa");
                }
            }
        }
        throw new SQLException("No hay tasa de cambio configurada de "
                + origen + " a " + destino + ".");
    }
}

File: CON BDD/EUREKABANK_REST_DOTNET_GR06/08. SERVIDOR/Data/TasaCambioDAO.cs
using Microsoft.Data.SqlClient;

namespace SERVIDOR.Data;

public class TasaCambioDAO
{
    public double Tasa(SqlConnection cn, string origen, string destino)
    {
        if (string.IsNullOrEmpty(origen) || string.IsNullOrEmpty(destino) || origen == destino)
            return 1.0;

        const string sql = """
            SELECT dec_tasa FROM tasacambio
            WHERE chr_origen = @origen AND chr_destino = @destino
            """;
        using var ps = new SqlCommand(sql, cn);
        ps.Parameters.AddWithValue("@origen", origen);
        ps.Parameters.AddWithValue("@destino", destino);
        using var rs = ps.ExecuteReader();
        if (rs.Read())
            return rs.GetDouble(0);

        throw new Exception($"No hay tasa de cambio configurada de {origen} a {destino}.");
    }
}
